cli/ballerina-cli/src/main/java/io/ballerina/cli/launcher/Main.java

```diff
@@ -83,7 +87,27 @@ private static Optional<BLauncherCmd> getInvokedCmd(String... args) {
             cmdParser.setStopAtPositional(true);
 
             // loading additional commands via SPI
-            ServiceLoader<BLauncherCmd> bCmds = ServiceLoader.load(BLauncherCmd.class);
+            ServiceLoader<BLauncherCmd> bCmds;
+            if (null != args && args.length > 0 && LauncherUtils.isToolCommand(args[0])) {
+                List<File> jars = LauncherUtils.getToolCommandJarAndDependenciesPath(args[0]);
+                URL[] urls = jars.stream()
+                        .map(file -> {
+                            try {
+                                return file.toURI().toURL();
+                            } catch (MalformedURLException e) {
+                                throw LauncherUtils.createUsageExceptionWithHelp("invalid tool jar : " + file
+                                        .getAbsolutePath());
+                            }
+                        })
+                        .toArray(URL[]::new);
+                // Combine custom class loader with system class loader
+                ClassLoader systemClassLoader = ClassLoader.getSystemClassLoader();
+                ClassLoader combinedClassLoader = new URLClassLoader(urls, systemClassLoader);
```

Comment:
If the jar is present in both classloaders, can we guarantee that the class in the URLClassloader is added to the classpath?

Reply:
Yes, the URLClassLoader prioratizes the urls over the system class loader.